### DISCUSSION TITLE
(FIXED) SOHO data fails to download (changed URL)

### DISCUSSION THREAD

**Discussion by Gandoon** [XEphem/XEphem#23] [Q&A] [answered]
Hello all,

This is an issue that has persisted for some time, and I have so far not been able to find a solution.
I currently run a macports v4.0.1 XEphem, and have previously been using v3.7.7 (non-macports).

The issue I have is that when opening the Sun view and trying to update to latest available image, the application throws an error due to a defunct URL.
This has been the case since at least v3.7.7
The original URL returns a '301 Moved Permanently' and the new location is given.

Is there any way of fixing this easily?
See log excerpt below.

Thanks :)

```
http: GET http://sohowww.nascom.nasa.gov/data/realtime/eit_284/1024/latest.jpg HTTP/1.0
User-Agent: xephem/4.0.1

Rcv: HTTP/1.1 301 Moved Permanently
Rcv: Date: Thu, 12 Aug 2021 15:41:05 GMT
Rcv: Server: Apache
Rcv: Strict-Transport-Security: max-age=31536000; includeSubdomains;
Rcv: Location: https://sohowww.nascom.nasa.gov/data/realtime/eit_284/1024/latest.jpg
Rcv: Content-Length: 277
Rcv: Connection: close
Rcv: Content-Type: text/html; charset=iso-8859-1
Rcv: <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
Rcv: <html><head>
Rcv: <title>301 Moved Permanently</title>
Rcv: </head><body>
Rcv: <h1>Moved Permanently</h1>
Rcv: <p>The document has moved <a href="https://sohowww.nascom.nasa.gov/data/realtime/eit_284/1024/latest.jpg">here</a>.</p>
Rcv: </body></html>
Error talking to SOHO .. see File->System log

**Comment by brandon-rhodes** [comment, marked as answer]
@Gandoon —

[**Update:** Sun images are now working in XEphem 4.1.0!]

Those resources have switched to "https", thus, SSL, and we don't have any C coders with security experience who can review the relevant patches.

I'll go take another look at them. I guess, if they have a buffer overflow, that the danger is that someone would take over one of the servers involved, like sohowww.nascom.nasa.gov, and return binary data that would exploit the overflow and make XEphem give them access to your cookies and passwords. So it would have to be a pretty specific attack. (Unless you were in a coffee shop and they could cleartext override the DNS addresses you were returned for servers like sohowww.nascom.nasa.gov? Hmm.)

Anyway, as no one else has stepped forward, I'll go do a Web search for how to vet SSL C code, and let you know what I find.

**Reply by Gandoon** [reply]
Thank you.
I had a look at the code a while back, when 3.7.7 was the latest and greatest version out there, and noticed the absence of SSL capabilities. Unfortunately I don't have the relevant know-how myself and no time to read up on it, so I dropped it for the time being. Here we are several years later and I just noticed the 4.x releases so I tested again if someone had managed to figure it out, but sadly no.
I thought I should ask if someone else had any input and there was something simple that I had missed.
Cheers

**Reply by brandon-rhodes** [reply]
Some good news — with the release of XEphem 4.1.0 back in September, Sun images are now downloading successfully again! I can open the Sun view, select "Download latest", and see a new image of the Sun appear. Let us know if you have any problems from now on with that feature!

**Reply by Gandoon** [reply]
That was a great effort… Many thanks :)

I will leave a few notes that popped up as it did not compile successfully straight out of the box this time (cloned directly from the github repository).
First, one needs to make sure the make-file points to the locations of libraries and includes if not in standard locations (i.e. if one uses packaging systems such as homebrew or MacPorts, keep this in mind), to make sure that the SSL libraries are found. That one was easy to solve. I uncommented and used the section already prepared for that.
Second, and this one was trickier: Neither the included Motif (at first) nor a homebuilt OpenMotif or Motif 2.3.8 seemed to result in an application that doesn't segfault. But I think that some of these issues may have been down to the xephem binary not being able to make the correct calls to start X on MacOS. I did eventually manage to get it working once I got it to compile properly using the included libXm and also making sure X was running before trying to start xephem. I have seen a few other comments regarding other builds of motif, but I cannot pinpoint what I am doing wrong, so I leave it at that, as it works now.
Third, and this is just a typo I guess: In skyfits.c, on line 1431 there is a cast to `(u_long)` in the section named `/* report progress */` that was styled `pmset ((ulong)fip->nbytes * 100 / fip->totbytes);`, after changing it to the former, it compiled as it should. However, if there is some override somewhere there may be something I missed (that specific cast is not present in the release tar bundle by the way).

Finally I can report that it builds just as fine with clang 13.0.0 as with gcc 11.2.0. No issues with either of them as far as I have noticed. The clang build turned out around 114 kB smaller than the gcc build :)

**Reply by brandon-rhodes** [reply]
> That was a great effort… Many thanks :)

Wow, you were able to compile and run under macOS? Over in https://github.com/XEphem/XEphem/discussions/13 folks have been trying for weeks without success — but maybe their install is different from yours?

> First, one needs to make sure the make-file points to the locations of libraries and includes if not in standard locations (i.e. if one uses packaging systems such as homebrew or MacPorts, keep this in mind), to make sure that the SSL libraries are found.

If your settings might be useful to other folks, feel free to share them, and I could add them as an alternative in the commented out options in the Makefile for macOS.

> Second, and this one was trickier: Neither the included Motif (at first) nor a homebuilt OpenMotif or Motif 2.3.8 seemed to result in an application that doesn't segfault. But I think that some of these issues may have been down to the xephem binary not being able to make the correct calls to start X on MacOS. I did eventually manage to get it working once I got it to compile properly using the included libXm and also making sure X was running before trying to start xephem. I have seen a few other comments regarding other builds of motif, but I cannot pinpoint what I am doing wrong, so I leave it at that, as it works now.

Were there any Makefile variables you changed? Or did it simply need to have X up and running?

> Third, and this is just a typo I guess: In skyfits.c, on line 1431 there is a cast to `(u_long)` in the section named `/* report progress */` that was styled `pmset ((ulong)fip->nbytes * 100 / fip->totbytes);`, after changing it to the former, it compiled as it should. However, if there is some override somewhere there may be something I missed (that specific cast is not present in the release tar bundle by the way).

Interesting — that line of code is relatively new. I wonder why "ulong"? Elsewhere in the code is `uLong` but not all lowercase `ulong`. I might simply change it to `int`, since that's the declared type of that function parameter.

Thanks for sharing your notes, and congratulations, again, on successfully getting it to work!

**Reply by Gandoon** [reply]
Wow, I thought I was just being punished for using a semi-antiquated MacOS… I am running v10.15.7 for compatibility reasons, but I use most other stuff as up to date as possible. I am running on x86_64, so maybe that makes things work more smoothly. I am not surprised if people have issues with the new ARM-based systems, but I thought the Intel machines should work reasonably well.

I did see the issues people had in the mentioned [#13](https://github.com/XEphem/XEphem/discussions/13), but I did have a 4.0.1 working that installed via MacPorts. Before that, I had a "homebuilt" (from tarball) 3.7.7 that worked just fine, so I persevered and played around until I got it to compile, and then I worried about the issue with X. The main take-home message would be that Motif is _not_ working very reliably on MacOS. But I guess most people know that by now…

As I have OpenSSL built with MacPorts, the way it worked for me was to gently edit the Makefile to the here attached form (turned out I didn't need the `ifndef MOTIFI` and `ifndef MOTIFL` edits, as I still had to pass the included motif with the suggested `make MOTIF=../../libXm/osx` anyway).
The only important thing was to make sure I got the /opt/local/lib and /opt/local/include paths passed properly so SSL could be found. 

And of course I changed the casting in skyfits.c as I mentioned to `(u_long)` and that works (and I guess it should be fine logic-wise, if I get any issues I will just remove it as I see that you already fixed it in the repository by removing it altogether).
I built it with clang v13 (and alternatively gcc 11) from MacPorts, so it is true that I haven't tried with the Apple-provided clang (/usr/bin/gcc links to the Apple provided clang, which is v12 on my system).

Finally, the trick to get it running was to manually start X (XQuartz 1.20.11, also from MacPorts), and then run xephem.

Cheers :)

(I had to add .txt as the system didn't like me just dropping the Makefile as is here.)
[Makefile.txt](https://github.com/XEphem/XEphem/files/7664383/Makefile.txt)
.

**Reply by Gandoon** [reply]
I marked the original reply as an answer to my question, as what came from this thread answered the question I had originally :)

**Reply by brandon-rhodes** [reply]
Thanks for replying there! Hopefully XEphem will soon be a more reliable compile experience on macOS.

**Reply by Gandoon** [reply]
Of course, no worries… It seemed as if it could be potentially useful to someone else with similar issues and configuration to mine.
Once again, thanks :)
